DOCSP-42794: Laravel Passport (#3113)
jenssegers/laravel-mongodb: docs/user-authentication.txt

   To learn more about the ``DocumentModel`` trait, see
   :ref:`laravel-third-party-model` in the Eloquent Model Class guide.

.. _laravel-user-auth-passport:

Laravel Passport
~~~~~~~~~~~~~~~~

Laravel Passport is an OAuth 2.0 server implementation that offers
API authentication for Laravel applications. Use Laravel Passport if

> your application requires OAuth2 support.

Reviewer: S: maybe you can link to a resource describing Oauth and/or passport?

to store OAuth2 clients, run the following command from your project root:

.. code-block:: bash

   php artisan install:api --passport

Next, navigate to your ``User`` model and add the ``Laravel\Passport\HasApiTokens``
trait. This trait provides helper methods that allow you to inspect a user's
authentication token and scopes. The following code shows how to add ``Laravel\Passport\HasApiTokens``
to your ``app\Models\User.php`` file:

.. code-block:: php

   <?php

   namespace App\Models;

   use MongoDB\Laravel\Auth\User as Authenticatable;
   use Laravel\Passport\HasApiTokens;

   class User extends Authenticatable
   {
      use HasApiTokens;
      ...
   }

Then, define an ``api`` authentication guard in your ``config\auth.php``
file and set the ``driver`` option to ``passport``. This instructs your
application to use Laravel Passport's ``TokenGuard`` class to authenticate
API requests. The following example adds the ``api`` authentication guard
to the ``guards`` array:

.. code-block:: php
   :emphasize-lines: 7-10

   'guards' => [
      'web' => [
         'driver' => 'session',
         'provider' => 'users',

>       ],

Reviewer: Is this other entry default?

Author: Yes the web entry is added automatically

>    

Reviewer: S: remove space (will need to adjust emphasized lines)

      'api' => [
         'driver' => 'passport',
         'provider' => 'users',
      ],
   ],

Use Laravel Passport with Laravel MongoDB
`````````````````````````````````````````

After installing Laravel Passport, you must enable Passport compatibility with MongoDB by
defining custom {+odm-short+} models that extend the corresponding Passport models.
To extend each Passport model class, include the ``DocumentModel`` trait in the custom models.
You can define the following {+odm-short+} model classes:

- ``MongoDB\Laravel\Passport\AuthCode``, which extends ``Laravel\Passport\AuthCode``
- ``MongoDB\Laravel\Passport\Client``, which extends ``Laravel\Passport\Client``
- ``MongoDB\Laravel\Passport\PersonalAccessClient``, which extends ``Laravel\Passport\PersonalAccessClient``
- ``MongoDB\Laravel\Passport\RefreshToken``, which extends ``Laravel\Passport\RefreshToken``
- ``MongoDB\Laravel\Passport\Token``, which extends ``Laravel\Passport\Token``

> For example, the following code overrides the default ``Laravel\Passport\AuthCode``
> model class by defining a ``MongoDB\Laravel\Passport\AuthCode`` class and including
> the ``DocumentModel`` trait:

Reviewer: S: possibly clearer explanation:
Something like
```suggestion
The following example code extends the default ``Laravel\Passport\AuthCode``
model class when defining a ``MongoDB\Laravel\Passport\AuthCode`` class and includes
the ``DocumentModel`` trait:
```

.. code-block:: php

   class MongoDB\Laravel\Passport\AuthCode extends Laravel\Passport\AuthCode
   {
      use MongoDB\Laravel\Eloquent\DocumentModel;

      protected $primaryKey = '_id';
      protected $keyType = 'string';
   }

> 

Reviewer: Q: does the user need to create models for each of the listed classes? If so, that should be more clear in a note or something similar.

Author: Yeah each one needs a custom model. I changed the wording of the paragraph below (line 246) to clarify that, lmk if it's clear enough

After defining each custom model, instruct Passport to use the models in the ``boot()`` 
method of your application's ``App\Providers\AppServiceProvider`` class, as shown
in the following code:

.. literalinclude:: /includes/auth/AppServiceProvider.php
   :language: php
   :emphasize-lines: 26-30
   :dedent:

> You can now use Laravel Passport and MongoDB in your application. For more information, see
> `Laravel Passport <https://laravel.com/docs/{+laravel-docs-version+}/passport>`__ in the
> Laravel documentation.

Reviewer: S: recommend moving the link up to the intro 
```suggestion
The, you can use Laravel Passport with MongoDB in your application.
```
